components/script/dom/element.rs

```diff
@@ -2702,20 +2785,30 @@ impl VirtualMethods for Element {
         }
     }
 
-    fn bind_to_tree(&self, tree_in_doc: bool) {
+    fn bind_to_tree(&self, tree_connected: bool) {
         if let Some(ref s) = self.super_type() {
-            s.bind_to_tree(tree_in_doc);
+            s.bind_to_tree(tree_connected);
         }
 
         if let Some(f) = self.as_maybe_form_control() {
             f.bind_form_control_to_tree();
         }
 
-        if !tree_in_doc {
+        if !tree_connected {
             return;
         }
 
         let doc = document_from_node(self);
+
+        if let Some(shadow_root) = self.upcast::<Node>().owner_shadow_root() {
+            let shadow_root = shadow_root.upcast::<Node>();
+            shadow_root.set_flag(NodeFlags::IS_CONNECTED, tree_connected);
```

Comment:
Touching the shadow root flags here looks wrong, we should do that in `element.rs` when binding the element instead, the same way you remove them.